### PR TITLE
Add cloudformation template and script for S3 bucket

### LINKED ISSUE #7: Write automation for the creation of S3 bucket to serve the React files
### User Story
As a developer, I want to write a script that creates S3 buckets, so I don't have to do it manually every time

### Estimated time to complete
4 hours

### Acceptance Criteria
- [x] Create script that automatically generates a new S3 bucket
    - [x] Script should accept a name parameter for the bucket

### Pull Request
#18

### PR DESCRIPTION
**Pull request by dansjack**
# Description
Closes #7 

This PR adds a [Cloudformation](https://aws.amazon.com/cloudformation/) template for provisioning an S3 bucket, and a shell script to deploy it. The bucket will eventually be used to serve a React app.

# Gallery

An example of the script successfully running 
image

If your bucket name isn't unique, this is the error message you'll get when you run `aws cloudformation describe-stack-events --stack-name <<STACK_NAME>>`
image

This shows the following:
1. navigating to the `devops` folder
2. attempting to run the command with a non-unique bucket name
3. reading the error logs
4. deleting the stack from the previous failed command `aws cloudformation delete-stack --stack-name <<STACK_NAME>>`
    * Note: `<<STACK_NAME>>` here begins with `arn` and ends with a string like `a36c6130-77f3-11ec-b26b-06300054abe1`
5. running the command successfully with a unique bucket name
image

Example page served from S3 bucket
image

# Testing

## Pre-reqs
* You need a `bash` or `zsh`. If you're on Windows, there are lots of guides for adding bash, [here's one](https://www.configserverfgithirewall.com/windows-10/install-bash-on-windows-10/)
* You must have the aws cli installed and it must be configured to use your access and secret key

## Steps
1. From the root of the repo, `cd devops/makeReactBucket`
2. Run the following: `./makeReactBucket.sh`
3. Enter your initials when prompted and press Enter
    * must only consist of letters a-z
    * can be no less than 2 chars
    * can be no more than 5 chars
4. You'll see a preview of your bucket and stack name. Confirm you want to create the stack and bucket by entering "y" and pressing Enter
    * The name will be formatted like this: `$initials-$randomString-YYYYMMDD-bucket`, `$initials-$randomString-YYYYMMDD-stack`
5. Once the stack is created, make sure the new stack and bucket have been created in the [S3](https://s3.console.aws.amazon.com/s3/home?region=us-west-2) and [Cloudformation](https://us-west-2.console.aws.amazon.com/cloudformation/home?region=us-west-2#/stacks?filteringStatus=active&filteringText=&viewNested=true&hideStacks=false) dashboards.
4. Copy/paste the following into a file called `index.html`:
```html
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Public site</title>
</head>
<body>
  <h1>Hello, world!</hi>
</body>
</html>

```
5. In the Objects tab of your new S3 bucket, click upload and then choose the file you just created, then click upload again
6. Go to the Properties tab of the S3 bucket and scroll to the bottom. Go to the url for your Bucket website endpoint. If everything worked, you should see the Hello, world! text like in the Gallery above.
7. Once done testing, make sure to empty and delete your S3 Bucket, then delete your cloudformation stack (in that order)

## Troubleshooting
I came across three issues during testing:
* Error when I gave a non-unique bucket name (solved by replacing with a unique name)
* Error when trying to update a stack that had failed to complete on the last attempt (solved by deleting the stack with `aws cloudformation delete-stack --stack-name`. See gallery screenshot 2 for more info
* You may need to run `chmod 755 makeReactBucket.sh` if you get an OS permissions error

### CODE REVIEW

File: devops/makeReactBucket.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Parameters": {
+    "BucketName": {
```
Comment (dansjack) [inline devops/makeReactBucket.json:4]: we set this with `--parameter-overrides BucketName=<<BUCKET_NAME>>` and get it with `{ "Ref": "BucketName" }` in the template below

**Review by KristinJJ — APPROVED** [review]
Looks Good to Me! Was able to successfully create an S3 bucket using the template following the testing steps. Steps for debugging were also helpful, as I also ran into the error of not choosing a unique bucket name the first time.

**Comment by dansjack** [comment]
Added a shell script to take the user's initials and deploy a new bucket and stack with unique resource names

**Review by aaronallenCS — APPROVED** [review]
Looks good to me!

**Review by Annelisebx — APPROVED** [review]

**Review by KristinJJ — APPROVED** [review]
Looks good to me! The script works and created a S3 bucket with my initials and a random string :)

**Pull request merged**